src/rust_option.rs

```diff
@@ -0,0 +1,527 @@
+impl<T: Sized> OptionTarget for Box<T> {}
+
+#[repr(C)]
+union OptionInner<T: OptionTarget> {
+    value: ManuallyDrop<T>,
+    empty: usize,
+}
+
+impl<T: OptionTarget> OptionInner<T> {
+    fn none() -> Self {
+        Self { empty: 0 }
+    }
+
+    fn new(t: T) -> Self {
+        Self { value: ManuallyDrop::new(t) }
+    }
+
+    fn has_value(&self) -> bool {
+        let _: () = assert_option_safe::<&T>();
+        unsafe { self.empty != 0 }
+    }
+
+    fn into_inner_unchecked(mut self) -> T {
+        let value = unsafe { ManuallyDrop::take(&mut self.value) };
+        unsafe { core::ptr::write(&mut self as _, OptionInner::none()) };
+        value
+    }
+}
+
+impl<T: OptionTarget> Drop for OptionInner<T> {
+    fn drop(&mut self) {
+        if self.has_value() {
+            unsafe { ManuallyDrop::drop(&mut self.value) };
+        }
+    }
+}
+
+// ABI compatible with C++ rust::Option<T> (not necessarily core::option::Option<T>).
+#[repr(C)]
+pub struct RustOption<T: OptionTarget> {
+    inner: OptionInner<T>,
+}
+
+pub const fn assert_option_safe<T>() {
+    struct __SizeCheck<U>(core::marker::PhantomData<U>);
+    impl<U> __SizeCheck<U> {
+        const _IS_NICHE: () =
+            assert!(core::mem::size_of::<Option<U>>() == core::mem::size_of::<U>());
+        const _IS_USIZE_SIZE: () =
+            assert!(core::mem::size_of::<Option<U>>() == core::mem::size_of::<usize>());
+    }
+    // Force the constants to resolve (at compile time)
+    let _: () = __SizeCheck::<T>::_IS_NICHE;
+    let _: () = __SizeCheck::<T>::_IS_USIZE_SIZE;
+}
+
+impl<T: OptionTarget> RustOption<T> {
+    pub fn new() -> Self {
+        let _: () = assert_option_safe::<&mut T>();
+        RustOption {
+            inner: OptionInner::none(),
+        }
+    }
+
+    pub fn value(&self) -> Option<&T> {
+        if self.has_value() {
+            unsafe { Some(self.inner.value.deref()) }
+        } else {
+            None
+        }
+    }
+
+    pub fn into_value(self) -> Option<T> {
+        if self.has_value() {
+            unsafe { Some(self.into_inner_unchecked()) }
+        } else {
+            None
+        }
+    }
+
+    pub fn has_value(&self) -> bool {
+        self.inner.has_value()
+    }
+
+    pub fn set(&mut self, value: T) {
+        self.inner = OptionInner {
+            value: ManuallyDrop::new(value),
+        }
+    }
+
+    pub unsafe fn into_inner_unchecked(self) -> T {
+        self.inner.into_inner_unchecked()
+    }
+
+    pub unsafe fn as_ref_inner_unchecked(&self) -> &T {
+        unsafe { self.inner.value.deref() }
+    }
+
+    pub unsafe fn as_ref_mut_inner_unchecked(&mut self) -> &mut T {
+        unsafe { self.inner.value.deref_mut() }
+    }
+}
+
+impl<'a, T> RustOption<&'a T>
+where
+    &'a T: OptionTarget,
+{
+    pub fn from_option_ref(other: Option<&'a T>) -> Self {
+        let _: () = assert_option_safe::<&T>();
+        unsafe { core::mem::transmute::<Option<&'a T>, RustOption<&'a T>>(other) }
```

Comment:
Earlier on your documented "ABI compatible with C++ rust::Option<T> (not necessarily core::option::Option<T>)." but here and throughout I see transmutations between `RustOption` and `Option`. Is the documentation wrong / imprecise?

Reply:
This documentation is the same as for Vec which does the same kind of transmutes. It was explained to me as the layout is the same for our RustOption and std::option::Option, but std::option::Option doesn't necessarily have the same calling convention as RustOption. That is, it could be passed in via registers instead of on the stack which would break the C++ side and is part of the ABI contract